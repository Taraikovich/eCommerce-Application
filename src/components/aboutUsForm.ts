

export class AboutUsForm {
  private form: HTMLDivElement;

  

  constructor() {
    this.form = document.createElement('div');
   
  }
  
  createForm(): HTMLDivElement {
    return this.form;
  }

}