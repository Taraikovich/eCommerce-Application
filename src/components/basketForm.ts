import {
  Cart,
  LineItem,
  LocalizedString,
  Image,
} from '@commercetools/platform-sdk';
import { getCart } from '../api/getCart';
import { changeLineItemQuantity } from '../api/changeLineItemQuantity';
import { removeFromCart } from '../api/removeFromCart';
import { createCart } from '../api/createCart';
import { clearCart } from '../api/clearCart';
import { Router } from '../router/router';
import { addDiscountCode } from '../api/addiscountCode';
import { getDiscountCode } from '../api/getDiscoutCode';

export class BasketForm {
  private form: HTMLDivElement;

  private itemsContainer: HTMLDivElement;

  callback?: () => void;

  private productsInCart: {
    id: string;
    key: string;
    name: LocalizedString;
    image: string;
    quantity: number;
    price: number;
    totalPrice: number;
  }[] = [];

  constructor() {
    this.form = document.createElement('div');
    this.itemsContainer = document.createElement('div');

    this.form.className = 'basket-form';
    this.itemsContainer.className = 'basket-items';

    this.form.appendChild(this.itemsContainer);
  }

  async createForm(callback?: () => void): Promise<HTMLDivElement> {
    this.callback = callback ? callback : this.callback;
    const cart: Cart = (await getCart()) as Cart;
    console.log(cart);
    this.productsInCart = [];
    this.itemsContainer.innerHTML = '';
    if (cart?.lineItems?.length) {
      cart.lineItems.forEach((item: LineItem) => {
        this.addItemToBasket({
          id: item.id,
          key: item.key as string,
          name: item.name,
          image: (item.variant?.images as Image[])[0]?.url,
          quantity: item.quantity,
          price: item.price.value.centAmount,
          totalPrice: item.totalPrice.centAmount,
        });
      });
      this.updateBasket();
      const clearCartButton = document.createElement('button');
      clearCartButton.textContent = 'Clear shoping cart';
      let totalCost = 0;
      for (const product of this.productsInCart) {
        totalCost += (product.price * product.quantity) / 100;
      }
      let totalCostWithDiscont = 0;
      for (const product of this.productsInCart) {
        totalCostWithDiscont += product.totalPrice / 100;
      }
      const totalCostElement = document.createElement('div');
      totalCostElement.className = 'total-cost';
      if (totalCost === totalCostWithDiscont) {
        totalCostElement.innerHTML = `<p>Total cost: ${totalCost.toFixed(
          2
        )} $</p>`;
      } else {
        totalCostElement.innerHTML = `<p>Total cost: ${totalCost.toFixed(
          2
        )} $</p>
                                      <p>With discont: ${totalCostWithDiscont.toFixed(
                                        2
                                      )} $</p>`;
      }

      clearCartButton.addEventListener('click', async () => {
        if (
          confirm("Do you want to clean out your customer's shopping cart?")
        ) {
          await clearCart();
          await createCart();
          await this.createForm();
          if (this.callback) {
            this.callback();
          }
        }
      });
      const discounCodes: string[] = [];
      cart.discountCodes.forEach((item) => {
        const code = item.discountCode.id;
        discounCodes.push(code);
      });
      console.log(discounCodes);

      this.itemsContainer.append(
        await this.discountCode(discounCodes),
        clearCartButton
      );
      this.itemsContainer.appendChild(totalCostElement);
    } else {
      const emptyCart = document.createElement('div');
      const emptyCartWrapper = document.createElement('div');
      emptyCartWrapper.className = 'empty-wrapper';
      emptyCart.className = 'empty-text';
      emptyCart.textContent = 'Your cart is empty!';
      const catalogLink = this.goToCatalog();
      emptyCartWrapper.appendChild(emptyCart);
      emptyCartWrapper.appendChild(catalogLink);
      this.itemsContainer.appendChild(emptyCartWrapper);
    }

    return this.form;
  }

  private goToCatalog(): HTMLButtonElement {
    const button = document.createElement('button');
    button.textContent = 'Go to Catalog';
    button.addEventListener('click', () => {
      const newURL = './products';
      window.history.pushState({}, '', newURL);
      document.body.textContent = '';
      new Router();
    });
    return button;
  }

  addItemToBasket(product: {
    id: string;
    key: string;
    name: LocalizedString;
    image: string;
    quantity: number;
    price: number;
    totalPrice: number;
  }) {
    this.productsInCart.push(product);
    console.log(this.productsInCart);
  }

  private updateBasket() {
    for (const product of this.productsInCart) {
      const leftButton = document.createElement('button');
      leftButton.textContent = '-';
      leftButton.addEventListener('click', async () => {
        await changeLineItemQuantity(product.id, product.quantity - 1);
        await this.createForm();
        if (this.callback) {
          this.callback();
        }
      });
      const quantityInCart = document.createElement('span');
      quantityInCart.textContent = `${product.quantity}`;
      const rightButton = document.createElement('button');
      rightButton.textContent = '+';
      rightButton.addEventListener('click', async () => {
        await changeLineItemQuantity(product.id, product.quantity + 1);
        await this.createForm();
        if (this.callback) {
          this.callback();
        }
      });
      const removeButton = document.createElement('button');
      removeButton.textContent = 'Remove from cart';
      removeButton.addEventListener('click', async () => {
        await removeFromCart(product.id);
        await this.createForm();
        if (this.callback) {
          this.callback();
        }
      });
      const item = document.createElement('div');
      item.className = 'item-cart';
      const itemWrapper = document.createElement('div');
      itemWrapper.className = 'item-cart-wrapper';
      const itemWrapperRight = document.createElement('div');
      itemWrapperRight.className = 'item-wrapper-right';
      const quantityWrapper = document.createElement('div');
      quantityWrapper.className = 'quantity-wrapper';
      item.textContent = product.name['en-US'];
      const image = new Image();
      image.src = product.image;
      image.width = 100;
      const price = document.createElement('div');
      console.log(product.totalPrice === product.price);
      if (product.totalPrice === product.price * product.quantity) {
        price.textContent = `${(product.price * product.quantity) / 100} $`;
      } else {
        price.innerHTML = `<span> ${
          (product.price * product.quantity) / 100
        } $</span> ${product.totalPrice / 100} $`;
      }

      item.appendChild(image);
      item.appendChild(price);
      itemWrapper.appendChild(item);
      quantityWrapper.appendChild(leftButton);
      quantityWrapper.appendChild(quantityInCart);
      quantityWrapper.appendChild(rightButton);
      itemWrapperRight.appendChild(quantityWrapper);
      itemWrapperRight.appendChild(removeButton);
      itemWrapper.appendChild(itemWrapperRight);
      this.itemsContainer.appendChild(itemWrapper);
    }
  }

  private async discountCode(cods: string[]): Promise<HTMLElement> {
    const wrapper = document.createElement('div');
    wrapper.className = 'discount-code';

    const input = document.createElement('input');
    input.className = 'discount-code__input';
    input.type = 'text';
    input.id = 'discount-code';
    input.placeholder = 'enter your code';

    const label = document.createElement('label');
    label.setAttribute('for', 'discount-code');
    label.textContent = 'Discount code: ';

    const message = document.createElement('p');
    if (cods.length) {
      async function getCodesNames(arr: string[]) {
        const names: string[] = [];
        for (let code of arr) {
          console.log(code);
          code = await getDiscountCode(code);
          names.push(code);
        }
        return names;
      }
      const codesNames = await getCodesNames(cods);
      message.textContent = `Promo codes applied: ${codesNames.join(', ')}`;
    } else {
      message.textContent = '';
    }

    const button = document.createElement('button');
    button.textContent = 'apply';
    button.className = 'discount-code__button';

    button.addEventListener('click', async () => {
      if (input.value) {
        if (await addDiscountCode(input.value)) {
          this.createForm();
        } else {
          message.textContent = `Discount code "${input.value}" not found`;
        }
      }
    });

    wrapper.append(label, input, button, message);

    return wrapper;
  }
}
