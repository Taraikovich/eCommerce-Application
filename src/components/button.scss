button {
  font-family: 'Tektur', cursive;
  background-color: #072326;
  color: #f2f2f2;
  padding: 5px;
  margin-left: 10px;
  border-radius: 5px;
  cursor: pointer;
  transition: all 0.4s ease;
  border: 1px solid #072326;
}
button:hover {
  background-color: #f2f2f2;
  color: #072326;
}

.button {
  font-family: 'Tektur', cursive;
  background-color: #072326;
  color: #f2f2f2;
  padding: 5px;
  margin-left: 10px;
  border-radius: 5px;
  cursor: pointer;
  transition: all 0.4s ease;
  border: 1px solid #072326;
}

.button:hover {
  background-color: #f2f2f2;
  color: #072326;
}

.button__add-to-cart {
  font-size: 20px;
  font-weight: bold;
  padding: 0 10px;
}

.button__add-to-cart:disabled {
  background-color: #c9f696;
  color: #072326;
  cursor: default;
}

.basket-button {
  width: 35px;
  height: 35px;
  cursor: pointer;
  padding-right: 40px;
}

.basket-button:hover {
  transform: scale(1.2);
}

.buttons-wrapper {
  display: flex;
}
