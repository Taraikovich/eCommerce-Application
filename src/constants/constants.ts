export const clientId = 'do8j3jCHTaaXS3QwfbHSiklY';
export const clientSecret = 'HWp7PMfiQmg2MBl44VudBm2GyW-iKen4';
export const projectKey = 'ecomerce-app';
export const scopes = [`manage_project:${projectKey}`];
