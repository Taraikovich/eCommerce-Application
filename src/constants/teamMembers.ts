export type PersonalInfo = {
  name: string;
  role: string;
  bio: string;
  github: string;
  featuresImplemented: string[];
};

export const igor: PersonalInfo = {
  name: 'Igor Taraikovich',
  role: 'Front-End Developer / Team leader',
  bio: 'Former System Administrator from Belarus. Currently live in Georgia.Holds a degree in Economics. Transitioning to the world of Frontend Development. Passionate about creating seamless user experiences and beautiful websites.',
  github: 'https://github.com/Taraikovich',
  featuresImplemented: [
    'webpack configuration',
    'router configuration',
    'api interaction',
    'basic site design',
    'catalog page',
    'product page',
    'discount code processing',
  ],
};

export const veronika: PersonalInfo = {
  name: 'Veronika Nemirova',
  role: 'Front-End Developer',
  bio: `I am eager to embark on my Front-End Development journey at RSSchool! Having successfully completed Stage 0 at RS School, I'm excited to further expand my knowledge and skills. I'm a quick learner who enjoys tackling various tasks, and I'm eager to gain fresh insights into the IT world.`,
  github: 'https://github.com/veronikanemirova',
  featuresImplemented: [
    'webpack configuration',
    'router configuration',
    'login form',
    'user page',
    'basket page',
  ],
};

export const alexey: PersonalInfo = {
  name: 'Alexey Petrov',
  role: 'Front-End Developer',
  bio: `Presently, I am engaged in freelance work based in Moscow, where my specialization is system administrator. My journey includes the successful completion of programming courses at Hexlet, providing me with a solid foundation in Python, Java, and JavaScript. My genuine eagerness is fueled by a desire to continuously explore and embrace new perspectives within the dynamic realm of IT.`,
  github: 'https://github.com/al-pet',
  featuresImplemented: ['webpack configuration', 'registration page'],
};

export const aboutUs = `<p>We are a trio of aspiring front-end developers, brought together by our shared journey at the RSSchool. Our dedication to achieving tangible results and our insatiable thirst for acquiring new skills have forged us into a formidable and close-knit team.</p>

<p>To kickstart our collaboration, we established a Discord group where we could seamlessly exchange vital information and host virtual meetings. Additionally, we leveraged GitHub Projects to efficiently manage our project, enabling us to break down complex tasks into manageable sub-tasks and distribute them among ourselves. This streamlined approach allowed us to maintain clarity in our project execution.</p>

<p>The remarkable synergy within our team, combined with our commitment to our shared goals, culminated in the creation of a successful product - the very website you are currently viewing on your screen.</p>`;
