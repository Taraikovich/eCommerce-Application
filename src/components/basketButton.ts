import { Router } from '../router/router';
import { Button } from './button';

export class BasketButton extends Button {
  constructor() {
    super();
    this.changePage();
  }

  createButton(): HTMLButtonElement {
    super.createButton();
    this.button.textContent = 'Basket';
    this.button.className = 'basket-button';
    return this.button;
  }

  private changePage(): void {
    this.button.addEventListener('click', () => {
      const newURL = '/basket';
      window.history.pushState({}, '', newURL);
      document.body.textContent = '';
      new Router();
    });
  }
}
