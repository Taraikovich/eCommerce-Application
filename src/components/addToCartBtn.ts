import { createApiBuilderFromCtpClient } from '@commercetools/platform-sdk';
import { Button } from './button';
import { client } from '../api/BuildClient';
import { projectKey } from '../constants/constants';
import { getCartId, getCartVersion } from '../state/getCart';
import { setCartVersion } from '../state/setCart';
import { addToCart } from '../api/addToCart';

export class AddToCartButton extends Button {
  constructor(id: string, key: string, productsInCart: string[]) {
    super();
    this.button.textContent = 'ðŸ›’ +';
    this.chengePage(id, key, productsInCart);
  }

  createButton(): HTMLButtonElement {
    super.createButton();
    this.button.className = 'button button__add-to-cart';
    return this.button;
  }

  private chengePage(
    productId: string,
    productKey: string,
    productsInCart: string[]
  ): void {
    if (productsInCart.includes(productId)) {
      this.button.textContent = 'ðŸ›’ âœ”';
      this.button.disabled = true;
    }

    this.button.addEventListener('click', async (event) => {
      event.stopPropagation();
      this.button.textContent = 'ðŸ›’ âœ”';
      this.button.disabled = true;
      await addToCart(productId, productKey)
    });
  }
}
