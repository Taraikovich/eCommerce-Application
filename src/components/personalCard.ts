import { PersonalInfo } from '../constants/teamMembers';

export class PersonalCard {
  createCard(data: PersonalInfo): HTMLElement {
    const wrapper = document.createElement('div');
    wrapper.className = 'personal-card';
    wrapper.append(
      this.addPhoto(data.name),
      this.addName(data.name),
      this.addRole(data.role),
      this.addBio(data.bio),
      this.addFeatures(data.featuresImplemented),
      this.addGithub(data.github)
    );
    return wrapper;
  }

  private addPhoto(name: string): HTMLElement {
    const photo = document.createElement('div');
    photo.className = `personal-card__photo personal-card__photo_${name}`;
    return photo;
  }

  private addName(personalName: string): HTMLHeadElement {
    const name = document.createElement('h3');
    name.textContent = personalName;
    return name;
  }

  private addRole(personalRole: string): HTMLParagraphElement {
    const role = document.createElement('p');
    role.textContent = personalRole;
    return role;
  }

  private addBio(personalBio: string): HTMLParagraphElement {
    const bio = document.createElement('p');
    bio.textContent = personalBio;
    return bio;
  }

  private addFeatures(features: string[]): HTMLElement {
    const listWrapper = document.createElement('div');

    const h3 = document.createElement('h3');
    h3.textContent = 'Contribution to this project:';

    const list = document.createElement('ul');
    features.forEach((item) => {
      const listItem = document.createElement('li');
      listItem.textContent = item;
      list.append(listItem);
    });

    listWrapper.append(h3, list);

    return listWrapper;
  }

  private addGithub(link: string): HTMLAnchorElement {
    const github = document.createElement('a');
    github.textContent = 'GitHub';
    github.target = '_blank';
    github.href = link;
    return github;
  }
}
