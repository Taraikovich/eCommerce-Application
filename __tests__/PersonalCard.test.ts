import { PersonalCard } from '../src/components/personalCard'; // Update with the correct path
import { PersonalInfo } from '../src/constants/teamMembers';

const personalCard = new PersonalCard();

describe('PersonalCard', () => {
  test('createCard should return an HTML element with the expected structure', () => {
    const testData: PersonalInfo = {
      name: 'John Doe',
      role: 'Software Engineer',
      bio: 'A software engineer passionate about coding.',
      featuresImplemented: ['Feature 1', 'Feature 2'],
      github: 'https://github.com/johndoe',
    };

    const card = personalCard.createCard(testData);

    expect(card).toMatchSnapshot();
  });
});
