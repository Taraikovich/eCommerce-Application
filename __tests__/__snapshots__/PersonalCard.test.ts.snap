// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersonalCard createCard should return an HTML element with the expected structure 1`] = `
<div
  class="personal-card"
>
  <div
    class="personal-card__photo personal-card__photo_John Doe"
  />
  <h3>
    John Doe
  </h3>
  <p>
    Software Engineer
  </p>
  <p>
    A software engineer passionate about coding.
  </p>
  <div>
    <h3>
      Contribution to this project:
    </h3>
    <ul>
      <li>
        Feature 1
      </li>
      <li>
        Feature 2
      </li>
    </ul>
  </div>
  <a
    href="https://github.com/johndoe"
    target="_blank"
  >
    GitHub
  </a>
</div>
`;
